### PR TITLE
Element form should prompt anytime you navigate away and there are changes pending

### LINKED ISSUE #512: Removing a behavior from the element form does not persist after saving
**Describe the bug**
When on the element form and clicking the ... menu and then "Remove behavior from element" appears to remove the behavior but when you click "Update element", you still see the behavior on the element both in the list and if you click on the element and go back to the form again.
NOTE: removing the element from the behavior (via behavior form) will work correctly

**To Reproduce**
1. Go to a scene with an element that has a behavior
2. Click on the Elements list tab
3. Click on an element which has a behavior
4. Click the "Behaviors" tab on the element form
5. Click the ... and select "remove behavior from element"
6. Click "Update Element"
7. See the "1 behavior" label in the elements list as well as the behavior if you go back into the form for the element.

**Expected behavior**
Should remove the behavior from the element 

**Screenshots**
Can't remove behavior from element

### LINKED ISSUE #572: Switch from using custom GUID generation to UUID package
**Is your feature request related to a problem? Please describe.**
Today we use some custom logic to generate GUIDs in the Utils.ts file. This is harder to mock and also, why should we be in this business?
image

**Describe the solution you'd like**
There is a library called UUID which has been used in all of my previous projects for GUID generation. We should be able to call their `v4` method that returns back a unique GUID.

Then the mocking is as simple as this
image

### PR DESCRIPTION
**Pull request by msnyder-msft**
### Summary of changes 🔍 
Migrating the state for the elements form to be in a reducer instead of in a single object. This allows us to track the dirty state of the object and know when changes have been made so that we can warn users when changes will be lost by navigating.

Moving all of the state management for the behavior form into a context. This allows us to track every update that is made to the behavior and know whether it has any changes or not. With this information, we can then show a confirmation dialog anytime the user tries to navigate away from the form. These flows include:
- cancel form
- mode switch toggle (viewer/builder)
- breadcrumb navigation

We do NOT show the prompt in any of those cases if no changes have been made OR if the changes were reverted back to matching the original state (ex: add an element and then remove it, then try to navigate, would not get a confirmation)

image

[Product Backlog Item 14504805](https://msazure.visualstudio.com/One/_workitems/edit/14504805): Dirty state prompt should be handled more holistically

Also fixes #512 
Fixes #572

### Testing 🧪
Making any changes on the element form and then trying to navigate somewhere else in the app should prompt you to confirm you want to discard changes. 

### Checklist ✔️
- [x] Linked associated issue (if present)
- [x] Added relevant labels
- [x] Requested developer reviews
- [x] Request UI review from design / PM
- [x] Verify all code checks are passing

### CODE REVIEW

File: src/Models/Context/ElementsFormContext/ElementFormContext.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+//     initialState: GET_MOCK_ELEMENT_FORM_PROVIDER_PROPS()
+// } as StoryProps;
+// RemoveAlias.play = async ({ canvasElement }) => {
+//     const canvas = within(canvasElement);
+//     // Finds the button and clicks it
+//     const elementsTabButton = await canvas.findByTestId(
+//         'ElementFormContext-RemoveLayer'
+//     );
+//     userEvent.click(elementsTabButton);
+// };
```
Comment (BrennaCourtney) [inline src/Models/Context/ElementsFormContext/ElementFormContext.stories.tsx:385]: block commenting to be deleted?

File: src/Models/Context/ElementsFormContext/ElementFormContextUtility.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    const hasBehaviorsChanged =
+        JSON.stringify(newBehaviors) !== JSON.stringify(originalBehaviorIds);
+
+    const isDirty = hasElementChanged || hasBehaviorsChanged;
+    logger(
+        'debug',
+        `IsFormDirty: ${isDirty}. ElementDirty: ${hasElementChanged}, BehaviorsDirty: ${hasBehaviorsChanged}`
+    );
+
+    return isDirty;
```
Comment (BrennaCourtney) [inline src/Models/Context/ElementsFormContext/ElementFormContextUtility.ts:32]: looks really cool! the logic is pretty easy to follow too I think

File: src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx [hunk 3]
```diff
@@ -143,259 +121,334 @@ const SceneElementForm: React.FC<IADT3DSceneBuilderElementFormProps> = ({
...
+                    ViewerConfigUtility.removeElementFromBehavior(
+                        formElementId,
                         behavior
                     );
+                    logDebugConsole(
+                        'debug',
+                        `Removing element ${formElementId} from behavior ${behavior.displayName} (${behavior.id}).`
+                    );
+
+                    // TODO: should we remove the behavior if it no long maps to any elements?
```
Comment (pasanchMSFT) [inline src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx:155]: Talk about this in adhoc dev time

File: src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx [hunk 4]
```diff
@@ -143,259 +121,334 @@ const SceneElementForm: React.FC<IADT3DSceneBuilderElementFormProps> = ({
...
+        onElementBackClick,
+        setUnsavedBehaviorChangesDialogOpen,
+        setUnsavedChangesDialogDiscardAction
+    ]);
+
     const handleCreateBehavior = useCallback(
         async (searchText: string) => {
             // Save element
-            await handleSaveElement();
+            handleSaveElement();
```
Comment (pasanchMSFT) [inline src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx:207]: Why is this no longer async?

File: src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx [hunk 5]
```diff
@@ -467,50 +489,59 @@ const BuilderLeftPanel: React.FC<IBuilderLeftPanelProps> = ({ styles }) => {
...
     );
 
+    const onDiscardChangesClick = useCallback(() => {
+        setUnsavedBehaviorChangesDialogOpen(false);
+        if (state.unsavedChangesDialogDiscardAction) {
+            state.unsavedChangesDialogDiscardAction();
+        }
+    }, [setUnsavedBehaviorChangesDialogOpen, state]);
+
     const onBreadcrumbNavigate = useCallback(
```
Comment (pasanchMSFT) [inline src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx:499]: walk-through this logic

File: src/Models/Context/BehaviorFormContext/BehaviorFormContext.stories.tsx [hunk 6]
```diff
@@ -264,115 +264,3 @@ ResetAfterChanges.play = async ({ canvasElement }) => {
     );
     userEvent.click(behaviorsTabButton);
 };
-
-// export const UpdateStorageUrl = Template.bind({});
```
Comment (pasanchMSFT) [inline src/Models/Context/BehaviorFormContext/BehaviorFormContext.stories.tsx:268]: Why were these here in the first place?

File: src/Models/Classes/ViewerConfigUtility.ts [hunk 7]
```diff
@@ -273,13 +273,21 @@ abstract class ViewerConfigUtility {
...
+     */
     static addBehaviorToScene(
         config: I3DScenesConfig,
         sceneId: string,
-        behavior: IBehavior
+        behavior: IBehavior,
+        updateInPlace = false
     ): I3DScenesConfig {
-        const updatedConfig = deepCopy(config);
+        const updatedConfig = updateInPlace ? config : deepCopy(config);
```
Comment (pasanchMSFT) [inline src/Models/Classes/ViewerConfigUtility.ts:290]: How does update in place help us? When do we want to do this?

File: .eslintrc.json [hunk 8]
```diff
@@ -12,9 +12,9 @@
         "prettier/@typescript-eslint"
     ],
     "rules": {
-        "no-debugger": "error",
+        "no-debugger": "warn",
         "no-multiple-empty-lines": "error",
-        "react-hooks/exhaustive-deps": "off",
```
Comment (pasanchMSFT) [inline .eslintrc.json:17]: I don't love not having exhaustive-deps but this should be off or all our PRs will have a ton of warnings

File: src/Models/Context/ElementsFormContext/ElementFormContext.stories.tsx (hunk 1 above)
Comment (msnyder-msft) [reply src/Models/Context/ElementsFormContext/ElementFormContext.stories.tsx:385]: Removed

File: src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx (hunk 3 above)
Comment (pasanchMSFT) [reply src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx:155]: Remove this todo

Comment (msnyder-msft) [reply src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx:155]: REmoved

File: src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx (hunk 4 above)
Comment (darsney) [reply src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx:207]: okay so this is actually a little scary, like if we create an element, we want to make sure that writing the config file succeeds BEFORE we go to the elements list, because the content of the elements list may be populated from reading the config file from storage

Comment (msnyder-msft) [reply src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx:205]: reverted

**Review by pasanchMSFT — APPROVED** [review]
LGTM

File: src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx [hunk 9]
```diff
@@ -179,94 +182,113 @@ const BuilderLeftPanel: React.FC<IBuilderLeftPanelProps> = ({ styles }) => {
...
+            dispatch({
+                type: SET_ADT_SCENE_BUILDER_SELECTED_ELEMENT,
+                payload: element
+            });
+        },
+        [dispatch]
+    );
+
+    const onCreateElementClick = useCallback(() => {
+        setSelectedElement(getDefaultElement({ id: v4() }));
```
Comment (darsney) [inline src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx:197]: I kind of liked using a utils function for this instead of v4 directly, kinda like your other PR that adds v4 to the utils.  do that plz prolly?

File: src/Models/Classes/3DVConfig.ts [hunk 10]
```diff
@@ -1,4 +1,5 @@
 import produce from 'immer';
+import { v4 } from 'uuid';
```
Comment (darsney) [inline src/Models/Classes/3DVConfig.ts:2]: no lint errors plz

File: src/Components/ADT3DSceneBuilder/Internal/Elements/Internal/AliasedTwinsTab.tsx [hunk 11]
```diff
@@ -1,4 +1,4 @@
-import React, { useCallback, useContext, useEffect, useState } from 'react';
+import React, { useCallback, useContext } from 'react';
```
Comment (darsney) [inline src/Components/ADT3DSceneBuilder/Internal/Elements/Internal/AliasedTwinsTab.tsx:1]: unused lint errors nooo

**Review by darsney — APPROVED** [review]
Haven't pulled it down or tested it, left a couple comments, +100 karma points for writing unit tests

File: src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx (hunk 9 above)
Comment (msnyder-msft) [reply src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx]: sure. done.
